src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs

```diff
@@ -0,0 +1,160 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var enumerableType))
+                {
+                    return;
+                }
+
+                var methodRuleDictionary = s_methodMetadataNames
+                    .SelectMany(m => enumerableType.GetMembers(m.MethodName)
+                                                   .OfType<IMethodSymbol>()
+                                                   .Where(method => method.TypeParameters.HasExactly(1)
+                                                                 && method.Parameters.HasExactly(1))
+                                                   .Select(method => (method, m.Rule)))
+                    .ToImmutableDictionary(key => key.method, v => v.Rule, SymbolEqualityComparer.Default);
+
+                if (methodRuleDictionary.IsEmpty)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context =>
+                {
+                    var invocation = (IInvocationOperation)context.Operation;
+
+                    var targetMethod = (invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod).OriginalDefinition;
+
+                    if (!methodRuleDictionary.TryGetValue(targetMethod, out var rule))
+                    {
+                        return;
+                    }
+
+                    var instanceArg = invocation.GetInstance(); // "this" argument of an extension method
+
+                    // because the type of the parameter is actually the non-generic IEnumerable,
+                    // we have to reach back through conversion operator(s) to get the "real" type of the argument
+                    if (instanceArg is not IConversionOperation conversionOperation)
+                    {
+                        return;
+                    }
+
+                    var argIEnumerableType = conversionOperation.Operand.Type.AllInterfaces.SingleOrDefault(t => t.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T);
+                    if (argIEnumerableType is null)
+                    {
+                        return;
+                    }
+
+                    var castFrom = argIEnumerableType.TypeArguments[0];
+                    var castTo = invocation.TargetMethod.TypeArguments[0];
+
+                    if (CastWillAlwaysFail(castFrom, castTo))
+                    {
+                        context.ReportDiagnostic(invocation.CreateDiagnostic(rule, castFrom.ToDisplayString(), castTo.ToDisplayString()));
+                    }
+                }, OperationKind.Invocation);
+            });
+
+            static bool CastWillAlwaysFail(ITypeSymbol castFrom, ITypeSymbol castTo)
```

Comment:
Hmm, I am wondering if we can use any compiler API here instead of trying to mimic conversion logic. @333fred Is there any compiler API that given two type symbols returns the `Conversion` info for it?

Reply:
One little/important detail here is that there are some subtle differences between the language  and IL casts. Eg. C# will emit calls to explicit and numeric conversion operators but the runtime will not. 

Note to self: ensure there is a test for this.

Reply:
I couldn't end up finding any API for this.